exonum/src/messages2/authorization.rs

```diff
@@ -0,0 +1,146 @@
+use failure::Error;
+use serde::Serialize;
+
+use crypto::{
+    self, hash, Hash, PublicKey, SecretKey, Signature, PUBLIC_KEY_LENGTH,
+    SIGNATURE_LENGTH,
+};
+use hex::{FromHex, ToHex};
+use messages::Message;
+
+use super::EMPTY_SIGNED_MESSAGE_SIZE;
+
+
+/// Correct raw message that was deserialized and verified, from `UncheckedBuffer`;
+/// inner data should be formed according to the following layout:
+/// | Position | Stored data |
+/// | - - - - - - - -| - - - - - - |
+/// | `0..32`  | author's PublicKey     |
+/// | `32`     | message class          |
+/// | `33`     | message type           |
+/// | `34..N`  | Payload                |
+/// | `N..N+64`| Signature                |
+///
+///
+/// Every creation of `SignedMessage` lead to signature verification, or data signing procedure,
+/// which can slowdown your code. Beware `SignedMessage` message, this procedure is not free.
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, Ord, PartialOrd)]
+pub struct SignedMessage {
+    pub(in messages) raw: Vec<u8>,
+}
+
+impl SignedMessage {
+    pub(crate) fn new(
+        cls: u8,
+        tag: u8,
+        value: Vec<u8>,
+        author: PublicKey,
+        secret_key: &SecretKey,
+    ) -> SignedMessage {
+        let mut buffer = Vec::new();
+        buffer.extend_from_slice(author.as_ref());
+        buffer.push(cls);
+        buffer.push(tag);
+        buffer.extend_from_slice(value.as_ref());
+        let signature = Self::sign(&buffer, secret_key).expect("Couldn't form signature");
+        buffer.extend_from_slice(signature.as_ref());
+        SignedMessage { raw: buffer }
+    }
+
+    /// Create `SignedMessage` wrapper from raw buffer.
+    /// Checks binary format and signature.
+    pub fn verify_buffer(buffer: Vec<u8>) -> Result<Self, Error> {
+        if buffer.len() <= EMPTY_SIGNED_MESSAGE_SIZE {
+            bail!("Message too short message_len = {}", buffer.len())
+        }
+        let signed = SignedMessage { raw: buffer };
+        let sign_idx = signed.raw.len() - SIGNATURE_LENGTH;
+
+        let pk = signed.author();
+        let signature = signed.signature();
+
+        Self::verify(&signed.raw[0..sign_idx], &signature, &pk)?;
```

Comment:
I guess we could make this an instance method of `SignedMessage` to avoid unpacking public key and signature here.